src/sim/mem_state.cc

```diff
@@ -107,26 +106,34 @@ MemState::isUnmapped(Addr start_addr, Addr length)
 void
 MemState::updateBrkRegion(Addr old_brk, Addr new_brk)
 {
-    /**
-     * To make this simple, avoid reducing the heap memory area if the
-     * new_brk point is less than the old_brk; this occurs when the heap is
-     * receding because the application has given back memory. The brk point
-     * is still tracked in the MemState class as an independent field so that
-     * it can be returned to the application; we just do not update the
-     * region unless we expand it out.
-     */
-    if (new_brk < old_brk) {
-        _brkPoint = new_brk;
-        return;
-    }
-
     /**
      * The regions must be page aligned but the break point can be set on
      * byte boundaries. Ensure that the restriction is maintained here by
      * extending the request out to the end of the page. (The roundUp
      * function will not round up an already aligned page.)
      */
-    auto page_aligned_brk = roundUp(new_brk, _pageBytes);
+    auto page_aligned_new_brk = roundUp(new_brk, _pageBytes);
+    auto page_aligned_old_brk = roundUp(old_brk, _pageBytes);
+
+    /**
+     * Reduce the heap memory area if the new_brk point is less than
+     * the * old_brk; this occurs when the heap is receding because the
+     * application has given back memory. This may involve unmapping
+     * heap pages, if new_brk rounds to a lower-address page. The
+     * previous, buggy behavior was to leave such pages mapped for
+     * simplicity. However, this introduced a bug where, if the heap
+     * was later expanded out, this routine reuses previously mapped
+     * heap pages that were not zero-filled (and instead were filled
+     * with stale application data), violating the assumptions of
+     * glibc malloc.
```

Comment:
Nit: I would reword this part of the comment as IMHO it's not entirely accurate to label previous behaviour as intrinsically buggy.

From what I understood not zeroing out the old heap is technically legal and it does not violate any glibc assumption as long as glibc is informed by such behaviour via the MORECORE_CLEAR value. We should just note that Linux expect the zeroing and unmapping pages anyway regardless of the MORECORE_CLEAR is the safest option that works for every system.